sdk/storage/azure-storage-file-share/azure/storage/fileshare/aio/_lease_async.py

```diff
@@ -46,22 +46,36 @@ class ShareLeaseClient(LeaseClientBase):
         A string representing the lease ID of an existing lease. This value does not
         need to be specified in order to acquire a new lease, or break one.
     """
+    def __init__(  # pylint: disable=missing-client-constructor-parameter-credential,missing-client-constructor-parameter-kwargs
+        self, client: Union["ShareFileClient", "ShareClient"],
+        lease_id: Optional[str] = None
+    ) -> None:
+        self.id = lease_id or str(uuid.uuid4())
+        self.last_modified = None
+        self.etag = None
+        if hasattr(client, 'file_name'):
+            self._client = client._client.file  # type: ignore # pylint: disable=protected-access
+            self._snapshot = None
+        elif hasattr(client, 'share_name'):
+            self._client = client._client.share
+            self._snapshot = client.snapshot
+        else:
+            raise TypeError("Lease must use ShareFileClient or ShareClient.")
 
     def __enter__(self):
         raise TypeError("Async lease must use 'async with'.")
 
-    def __exit__(self, *args):
+    def __exit__(self, *args: Any):
         self.release()
```

Comment:
I'll start the conversation with @annatisch. In terms of `__exit__`, should we make that throw an error or just directly remove both of them? I'm leaning on the side of making both throw errors since users should be using `async for` and not `for`, etc.

Reply:
For now, I'll update this `__exit__` by having it throw a `TypeError` exception, and leaving `__enter__` as is. This seems to be the pattern in the codebase. Let's definitely follow-up here tho.

Reply:
Remove `__enter__` and `__exit__`.